[Step1] Hailey

eunseo5355 opened the pull request:
✨@seizze✨

하이디 안녕하세요~!☺️
드디어 첫번째 PR를 보내게 되었네요😅
먼저 작동 화면부터 보여드릴게요! 

## 작동 화면
![화면 기록 2021-06-07 오후 6 05 34](https://user-images.githubusercontent.com/55755686/120991907-09acd480-c7bd-11eb-8160-f378c692047a.gif)
![화면 기록 2021-06-07 오후 5 56 35](https://user-images.githubusercontent.com/55755686/120992216-58f30500-c7bd-11eb-8ce2-e604a495f48b.gif)
![화면 기록 2021-06-07 오후 6 04 11](https://user-images.githubusercontent.com/55755686/120992470-98215600-c7bd-11eb-926d-e1109085528a.gif)


## 구현사항

### `Model`
 - `Note` : Note 데이터를 받아오기 위한 Model 타입

### `ViewModel`
- `NoteListViewModel` : Note 리스트 즉, 테이블 뷰에 필요한 데이터 모델
- `NoteViewModel` : Note 각 요소의 데이터에 접근하기 위한 데이터 모델

### `Delegate`
- `NoteDelegate` : table view에서 cell을 선택했을때 선택한 cell의 데이터를 전달해주기 위한 delegate

### `Error`
- `DataError` : 에셋에서 데이터를 가져오고 디코딩할때 발생하는 에러 타입 구현

### `ViewController`
- `NoteListTableViewCell` : 테이블 뷰에 필요한 셀 구현
- `NoteListViewController` 
  - NoteListViewModel을 통해 Model에 접근하여 Table View 구현
  - delegate를 이용하여 cell의 데이터를 전달
- `NoteDetailViewController`
  - Text View에 padding을 주기 위해 contentInset 사용
  - contentOffset을 사용하여 스크롤 위치 조절
  - 가로가 compact 사이즈일 때는 Text View의 배경색을 systemGray3 색상으로, regular 사이즈일 때는 systemBackground 색상으로 설정되도록 구현
  - `traitCollectionDidChange(_ previousTraitCollection:)` 메소드를 사용하여 화면의 가로가 compact였다가 기기를 가로모드로 돌렸을때 즉, regular로 바뀔때 Text View의 배경색이 systemBackground로 변경되도록 구현

- `SplitViewController` 
  - `NoteListViewController` 와  `NoteDetailViewController` 를 UINavigationController로 각각 만들어준 후 viewControllers에  넣어줌
  - NoteDelegate를 채택하여 받은 데이터를 NoteDetailViewController의 Text View에 넣어줌
  - UISplitViewControllerDelegate의 메소드를 사용하여 Split View가 축소되었을때 Secondary ViewController 즉, detailViewController가 listViewController에 통합되도록 구현하여 처음에 뜨는 화면을 listViewController로 설정

## 조언을 얻고 싶은 부분

1. Note의 날짜 포맷을 사용자의 지역에 따라 변하도록 구현했습니다!
잘 되는지 확인하기 위해 시뮬레이터의 지역을 바꿔보려고 했는데 아직 해결하지 못했어요😢 
지금 시뮬레이터는 미국으로 설정되어있는것 같아서 한국으로 바꿔주기 위해 시뮬레이터의 지역도 변경해보았고 좌표도 변경해봤는데 바뀌지 않더라구요,,ㅠ 혹시 하이디는 아실까 해서 여쭤보았습니다..!
<br>

2. 처음으로 MVVM 패턴을 구현해보았어요..!ㅎㅎ View Controller가 Model에 직접 접근하지 않도록 구현했는데 혹시 잘못된 부분이나 조언해주고 싶으신 부분이 있다면 조언 부탁드려요😃 
처음 구현해봐서 이렇게 하는게 맞는지 잘 모르겠네요,,😂
<br>

3. Cell의 UI를 Stack View를 사용하여 구현하였는데 스택뷰의 이름 짓기가 어려웠습니다..ㅠ
지금 네이밍이 괜찮은지 조언 얻고 싶어요..! 그리고 혹시 네이밍 팁 같은게 있을까요?🤩
<br>

4. 날짜 포맷해주는 과정이 NoteViewModel 안에 구현되어있는데 혹시 하이디는 Model과 ViewModel 중에 어디에 있는게 낫다고 생각하시나요??🤔 저는 Model은 정말 데이터 타입만을 가지고 있는게 나을 것 같다고 생각하여 View Model 타입에 넣었습니다!

리뷰 감사해요 하이디🥰

eunseo5355 committed:
✨ [Feat] 프로젝트 초기 설정

- ViewController -> SplitViewController로 이름 변경 후 UISplitViewController로 변경
- 스토리보드 삭제 후 SceneDelegate에서 초기 화면을 SplitViewController로 설정

eunseo5355 committed:
🍱 [Asset] 데이터 에셋 추가

eunseo5355 committed:
✨ [Feat] Model 구현

eunseo5355 committed:
✏️ [Docs] 주석 수정

eunseo5355 committed:
✨ [Feat] Model 타입에서 날짜 포맷해주는 프로퍼티 추가

- 구조체명 NoteItem -> NoteData로 수정

eunseo5355 committed:
✏️ [Docs] 주석 수정

eunseo5355 committed:
✨ [Feat] 테이블뷰 셀 구현

eunseo5355 committed:
♻️ [Refactor] dateFormat 형식 변경

eunseo5355 committed:
✨ [Feat] NoteDelegate 프로토콜 구현

eunseo5355 committed:
✨ [Feat] 메모 리스트뷰 구현

eunseo5355 committed:
✨ [Feat] 메모 상세화면 구현

eunseo5355 committed:
✨ [Feat] Split View 구현

- 가로가 compact size인 화면에서는 메모 리스트가 뜨도록 구현

eunseo5355 committed:
✏️ [Docs] 주석 수정

eunseo5355 committed:
✨ [Feat] 사용자의 지역에 따라 날짜 포맷되도록 수정

eunseo5355 committed:
♻️ [Refactor] 접근제어자 변경

eunseo5355 committed:
✨ [Feat] 가로 Compact Size에 따라 배경색이 변하도록 구현

eunseo5355 committed:
♻️ [Refactor] MVVM 패턴 적용

- Data Error 타입 구현
- NoteListViewModel, NoteViewModel 타입 구현
- ViewController가 Model에 직접 접근하던 것을 ViewModel을 통해 접근하도록 수정
- Moedel에 있던 formattedLastModified 프로퍼티를 ViewModel로 옯김

eunseo5355 committed:
✏️ [Docs] 필요없는 주석 삭제

eunseo5355 committed:
🎨 [Style] 필요없는 줄바꿈 삭제

eunseo5355 committed:
🚚 [Chore] 프로젝트 파일 정리

eunseo5355 committed:
♻️ [Refactor] NoteData → Note로 구조체명 변경

eunseo5355 committed:
🔨 [Chore] Info.plist 파일 위치 수정

eunseo5355 committed:
💄 [UI] textView 제약 수정

seizze commented on CloudNotes/CloudNotes/ViewController/NoteListViewController.swift:
```diff
@@ -0,0 +1,71 @@
+final class NoteListViewController: UIViewController {
+    private let noteTableView: UITableView = UITableView()
+    private var noteListViewModel: NoteListViewModel = NoteListViewModel()
+    static weak var noteDelegate: NoteDelegate?
```
delegate를 static으로 선언해도 괜찮을까요? 이 프로퍼티로 가리키고 있는 객체가 메모리에서 언제 해제될까요? ㅎㅎ

eunseo5355 replied:
지금처럼 static으로 선언하게 되면 해당 프로퍼티가 가리키고 있는 객체가 앱이 종료될때 메모리에서 해제되서 필요없을때도 메모리 공간을 차지하고 있을 것 같아요..! 따라서 noteDelegate가 필요한 클래스에 인스턴스를 생성한 후 delegate를 연결시켜주면 이 인스턴스가 필요없어질때 ARC가 자동으로 메모리에서 해제해줘서 더 효율적일 것 같다는 생각이 들었습니다! 혹시 제가 잘못 이해한 부분이 있다면 조언부탁드려요☺️

seizze replied:
훌륭합니다👏👏 해당 코드 push되면 한 번 더 체크해 볼게요~

seizze commented on CloudNotes/CloudNotes/ViewController/SplitViewController.swift:
```diff
@@ -0,0 +1,35 @@
+extension SplitViewController: NoteDelegate {
+    func showDetailNote(data: NoteViewModel) {
+        let noteDetailNC = UINavigationController(rootViewController: NoteDetailViewController())
+        (noteDetailNC.viewControllers.last as? NoteDetailViewController)?.configureTextView(data)
```
이런 식으로 NoteDetailViewController를 생성할 때마다 외부에서 configureTextView를 호출해주는 것보다, 데이터만 넘기고 스스로 세팅되도록 만드는 편이 좋습니다. 텍스트뷰 세팅은 NoteDetailViewController의 역할이 더 적합하기도 하고, 다른 사람이 앱의 다른 곳에서 NoteDetailViewController를 띄우게 될 때 configureTextView를 호출해서 설정해줘야 한다는 걸 인지하기 힘들 수 있으니까요~

eunseo5355 replied:
`NoteDetailViewController`에 데이터를 넘겨받을 프로퍼티를 두고 외부에서 데이터를 받아서 프로퍼티의 값이 변경되었을 경우 `configureTextView`가 호출되도록 수정했습니다!

seizze commented on CloudNotes/CloudNotes/ViewController/SplitViewController.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+extension SplitViewController: NoteDelegate {
+    func showDetailNote(data: NoteViewModel) {
+        let noteDetailNC = UINavigationController(rootViewController: NoteDetailViewController())
+        (noteDetailNC.viewControllers.last as? NoteDetailViewController)?.configureTextView(data)
+        self.showDetailViewController(noteDetailNC, sender: self)
+    }
```
근데 root view controller로 넘긴 NoteDetailViewController()와 `noteDetailNC.viewControllers.last`로 접근한 인스턴스가 동일하지 않나요? 취향 차이일 수도 있겠지만 저는 이렇게 하는게 가독성 면에서 더 좋은 것 같아요. ㅎㅎ

```swift
let noteDetailViewController = NoteDetailViewController()
noteDetailViewController.configureTextView(data)

let navigationController = UINavigationController(rootViewController: noteDetailViewController)
showDetailViewController(navigationController, sender: self)
```

eunseo5355 replied:
저도 이게 더 가독성이 좋아보이네요..!!👍

seizze commented on CloudNotes/CloudNotes/ViewModel/NoteListViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+import UIKit
+
+final class NoteListViewModel {
+    lazy var notes: [Note] = {
+        do {
+            return try self.decode()
+        } catch let error {
+            print(error.localizedDescription)
+            return []
+        }
+    }()
```
lazy로 선언된 저장 프로퍼티는 멀티 스레드 환경에서 안전하지 않을 수 있습니다. 참고 — [Stored Properties](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#ID255)
꼭 lazy로 선언할 필요가 없다면 생성자에서 지정해 주는 방식도 고려해 보세요 ㅎㅎ

eunseo5355 replied:
해당 클래스 안에 있는 `decode()` 메소드를 사용하기 위해 lazy로 선언했던 방식을 
init()을 사용하여  생성자 안에서 디코딩 작업과 `notes`를 지정해주는 작업을 하도록 수정하였습니다!

seizze commented on CloudNotes/CloudNotes/ViewController/Cell/NoteListTableViewCell.swift:
```diff
@@ -0,0 +1,79 @@
+        return label
+    }()
+    
+    private lazy var innerStackView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [dateLabel, previewBodyLabel])
+        stackView.distribution = .equalSpacing
+        stackView.axis = .horizontal
+        stackView.spacing = 20
+        return stackView
+    }()
+    
+    private lazy var outerStackView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [titleLabel, innerStackView])
+        stackView.distribution = .equalSpacing
+        stackView.axis = .vertical
+        stackView.spacing = 3
+        return stackView
+    }()
```
> Cell의 UI를 Stack View를 사용하여 구현하였는데 스택뷰의 이름 짓기가 어려웠습니다..ㅠ
지금 네이밍이 괜찮은지 조언 얻고 싶어요..! 그리고 혹시 네이밍 팁 같은게 있을까요?🤩

많이 지어봐서 고민하는 습관을 들이면 좋다고 생각합니다~! 저는 이름지을 대상들을 포괄하는 (추상적인) 다른 대상을 떠올리려 하면 도움이 되더라고요 ☺️
제가 이름짓는 방식대로 지어보자면, descriptionStackView, noteStackView 정도가 되겠네요. 하지만 innerStackView, outerStackView도 이해하는 데 큰 무리가 없기 때문에 괜찮다고 생각합니다. ㅎㅎ

seizze commented on CloudNotes/CloudNotes/ViewController/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        textView.textAlignment = .left
+        textView.font = UIFont.preferredFont(forTextStyle: .body)
+        setBackGroundColor(of: textView)
+        textView.contentOffset = CGPoint(x: 0, y: -20)
+        textView.contentInset = UIEdgeInsets(top: 20, left: 20, bottom: 20, right: 20)
```
contentOffset과 contentInset을 적절하게 사용하셨네요! 💯💯

seizze commented on CloudNotes/CloudNotes/ViewModel/NoteViewModel.swift:
```diff
@@ -0,0 +1,29 @@
+        return note.title
+    }
+    
+    var formattedLastModified: String {
+        let date = Date(timeIntervalSince1970: TimeInterval(note.lastModified))
+        let dateFormatter = DateFormatter()
+        dateFormatter.locale = Locale.current
+        dateFormatter.dateStyle = .medium
+        
+        return dateFormatter.string(from: date)
+    }
```
`Locale.current`를 찍어보니 "en_KR"로 나오고 있네요! 언어 설정이 en이라 날짜가 영어로 나왔을 것 같습니다.

1. `Locale.preferredLanguages`를 이용하거나
2. 프로젝트 설정에서 Korean Localization을 추가하는 방법이 있을 것 같아요.

[이 글](https://medium.com/swlh/know-your-language-locale-in-swift-beae4fcc5174)에 잘 정리되어 있네요~ ☺️

seizze reviewed (changes requested):
안녕하세요 Hailey~ 수고 많으셨습니다~!
아래에 코멘트에 남겨주신 내용에 관해 의견 남겼습니다. ✨

> Note의 날짜 포맷을 사용자의 지역에 따라 변하도록 구현했습니다!
잘 되는지 확인하기 위해 시뮬레이터의 지역을 바꿔보려고 했는데 아직 해결하지 못했어요😢
지금 시뮬레이터는 미국으로 설정되어있는것 같아서 한국으로 바꿔주기 위해 시뮬레이터의 지역도 변경해보았고 좌표도 변경해봤는데 바뀌지 않더라구요,,ㅠ 혹시 하이디는 아실까 해서 여쭤보았습니다..!

아래 코드에 의견 남겼습니다~!

> 처음으로 MVVM 패턴을 구현해보았어요..!ㅎㅎ View Controller가 Model에 직접 접근하지 않도록 구현했는데 혹시 잘못된 부분이나 조언해주고 싶으신 부분이 있다면 조언 부탁드려요😃
처음 구현해봐서 이렇게 하는게 맞는지 잘 모르겠네요,,😂

MVVM을 시도해 보셨네요! 😎
지금은 MVVM이라기보다, 좀더 객체지향적인 MVC 느낌입니다. ViewModel은 단순히 모델을 설명하는 역할이 아니고, 모델이 변경되었을 때 알려주는 기능이 있어야 합니다. 그리고 ViewModel과 View 사이는 바인딩으로 연결되어야 합니다. 이 글을 참조해서 리팩토링을 해 보시면 좋은 경험이 될 것 같아요. [iOS MVVM Tutorial: Refactoring from MVC](https://www.raywenderlich.com/6733535-ios-mvvm-tutorial-refactoring-from-mvc)

> Cell의 UI를 Stack View를 사용하여 구현하였는데 스택뷰의 이름 짓기가 어려웠습니다..ㅠ
지금 네이밍이 괜찮은지 조언 얻고 싶어요..! 그리고 혹시 네이밍 팁 같은게 있을까요?🤩

이 내용도 해당 코드에 의견 남겼습니다 ㅎㅎ

> 날짜 포맷해주는 과정이 NoteViewModel 안에 구현되어있는데 혹시 하이디는 Model과 ViewModel 중에 어디에 있는게 낫다고 생각하시나요??🤔 저는 Model은 정말 데이터 타입만을 가지고 있는게 나을 것 같다고 생각하여 View Model 타입에 넣었습니다!

객체지향과 MVVM은 동떨어진 개념이 아닙니다. 따라서 모델도 객체지향적으로 동작할 수 있어요. MVVM을 한다고 해서 모든 로직이 뷰모델에 존재해야 하는것은 아닙니다.
개인적인 의견으로 날짜 포맷팅 기능은, 보통 한 화면에서만 사용되지 않고 많은 곳에서 사용되기 때문에 앱 전반적으로 재사용될 수 있도록 유틸리티 형식으로 만드는 것도 좋다고 생각합니다. ☺️

이 외에도 코드에 피드백 남겼으니 참고 부탁드릴게요~!

eunseo5355 committed:
♻️ [Refactor] static이었던 noteDelegate 프로퍼티를 인스턴스 프로퍼티로 수정

eunseo5355 committed:
✨ [Feat] 프로젝트 Localization에 Korean 추가

eunseo5355 committed:
♻️ [Refactor] NoteDetailVC의 Text View가 데이터를 받아서 스스로 세팅되도록 수정

eunseo5355 committed:
♻️ [Refactor] 생성자에서 notes 프로퍼티에 데이터를 넣어주는 방식으로 수정

eunseo5355 committed:
♻️ [Refactor] MVVM 패턴 수정

- Observable 타입 생성
- ViewModel과 View 사이를 바인딩으로 연결시켜줌

eunseo5355 committed:
♻️ [Refactor] StackView 프로퍼티명 수정

eunseo5355 committed:
♻️ [Refactor] configureTextView() 호출 위치 변경

eunseo5355 commented:
@seizze 

안녕하세요 하이디~!
피드백 주신 내용을 바탕으로 수정해보았어요 :)

MVVM 바인딩이 좀 많이 어렵네요..😂
글 읽고 리팩토링 해 보았는데 하이디의 피드백이 필요합니다..!

이번에도 수정해야될 부분이 있다면 조언 부탁드려요 ㅎㅎ

seizze commented on CloudNotes/CloudNotes/Delegate/NoteDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol NoteDelegate: AnyObject {
+    func showDetailNote(data: NoteViewModel)
+}
```
delegate 패턴을 사용할 때는, 첫 번째 프로퍼티로 자기 자신을 넘겨 주는 게 좋습니다. 구현하는 쪽에서 여러 델리게이트들을 동시에 받아 처리할 경우 어떤 인스턴스에서 이벤트가 왔는지 판별하기 위해서요. 테이블뷰 delegate 메소드들을 구현할 때 첫 번째 프로퍼티로 어떤 게 넘어오는지를 보면 이해하기 쉬울 거에요~

eunseo5355 replied:
```Swift
protocol NoteDelegate: AnyObject {
    func showDetailNote(_ listViewController: NoteListViewController, data: NoteViewModel)
}
```
이렇게 수정해 보았어요..!

seizze commented on CloudNotes/CloudNotes/ViewController/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,67 @@
+import UIKit
+
+final class NoteDetailViewController: UIViewController {
+    var note: NoteViewModel? {
```
ViewModel을 참조하는 프로퍼티의 경우 명시적으로 ~뷰모델이라고 네이밍하는 것이 좋습니다. ㅎㅎ

seizze commented on CloudNotes/CloudNotes/Model/Note.swift:
```diff
@@ -0,0 +1,27 @@
+    let title: String
+    let body: String
+    let lastModified: Int
+    var formattedLastModified: String {
+        let date = Date(timeIntervalSince1970: TimeInterval(lastModified))
+        let dateFormatter = DateFormatter()
+        dateFormatter.locale = Locale.current
+        dateFormatter.dateStyle = .medium
+
+        return dateFormatter.string(from: date)
+    }
```
[Data Formatting Guide > Cache Formatters for Efficiency](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DataFormatting/Articles/dfDateFormatting10_4.html#//apple_ref/doc/uid/TP40002369-SW3)를 보시면, 아래와 같은 내용이 있어요.

> Creating a date formatter is not a cheap operation. If you are likely to use a formatter frequently, it is typically more efficient to cache a single instance than to create and dispose of multiple instances. One approach is to use a static variable.

DateFormatter를 생성하는 비용은 저렴하지 않기 때문에, 매번 생성하는 것 보다는 static 프로퍼티 등을 활용해서 한 번만 생성해 재사용하는 편이 효율적입니다. 🙂

eunseo5355 replied:
저번 코멘트에서 `유틸리티 형식` 이 부분을 이해 못해서 다시 여쭤보려고 했는데 까먹었네요😅 
다시 피드백 주셔서 감사해요😃
```Swift
static let dateFormatter: DateFormatter = {
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale.current
        dateFormatter.dateStyle = .medium
        return dateFormatter
    }()
```
이렇게 수정해 보았는데 하이디가 의도하신 바가 맞을까요?

seizze replied:
맞습니다~ 굿굿 💯

seizze commented on CloudNotes/CloudNotes/ViewController/NoteListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        configureNavigationItem()
+        configureConstraint()
+        
+        noteListViewModel.notes.bind { [weak self] _ in
+            DispatchQueue.main.async {
+                self?.noteTableView.reloadData()
+            }
+        }
```
잘 하셨네요~ ㅎㅎ 네트워크에서 새로운 데이터가 오든, 사용자가 삭제하든간에 상관없이 노트가 변경되면, 하나의 흐름으로 뷰가 다시 그려진다는 게 중요합니다. 스텝2의 CRUD 진행할 때 참고하시면 좋을 것 같아요. ☺️

seizze commented on CloudNotes/CloudNotes/Utility/Observable.swift:
```diff
@@ -0,0 +1,28 @@
+
+import Foundation
+
+final class Observable<T> {
```
오 여긴 어떤 이유로 Observable로 짓게 됐나요?

eunseo5355 replied:
관찰한다는 의미에서 Observable로 짓게 되었어요
즉, 쓸 때 `이 프로퍼티는 관찰 가능하다` 이런 식으로 해석하기 위해서요..!

seizze commented on CloudNotes/CloudNotes/ViewController/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,67 @@
+        ])
+    }
+    
+    func configureTextView() {
+        guard let note = note else { return }
+        self.detailNoteTextView.text = note.title + "\n\n" + note.body
+    }
```
1. 텍스트"뷰"에 데이터를 어떻게 표시할 지 만드는 로직은, 뷰모델의 역할이 더 적합해 보여요. 뷰는 테스트가 어렵기 때문에, 이런 로직 없이 단순히 데이터만 받아서 세팅하도록 만들고, 테스트가 가능한 뷰모델에서 어떤 데이터를 표시할 지 만들어 전달해 주는게 좋을 것 같습니다.

2. 뷰를 설정하는 코드는 viewDidLoad 또는 그 이후에 실행시키는 것이 좋습니다. 지금은 코드로 만든 데다가 스플릿뷰에서 VC들을 만들어 놓고 재사용하고 있어서 문제가 없지만, 스토리보드를 이용할 땐 viewDidLoad 이전 시점(예를 들면 인스턴스 생성 시점)은 뷰가 완전히 생성된 시점이 아니기 때문에 이 코드가 제대로 실행이 되지 않을 수 있어요. ㅎㅎ 코드는 변경에 유연하게 작성하는 편이 좋다고 생각합니다.

따라서 ViewModel은 지금처럼 외부에서 생성해 넣어주고, NoteListViewController처럼 viewDidLoad에서 텍스트뷰와의 바인딩 코드를 넣어보면 어떨까요?

eunseo5355 replied:
NoteDetailViewController에서 썼던 NoteViewModel에는 Observable 타입이 없어서 따로 NoteTextViewModel 타입을 새로 만들어주었습니다. 

1. NoteTextViewModel에 Text View가 어떤 데이터를 표시할지 만들어서 전달해주는 메소드를 구현했습니다

2. viewWillAppear() 안에 뷰를 설정하는 코드를 넣어주었어요

마지막으로 viewDidLoad에 뷰모델과 텍스트뷰와의 바인딩 코드를 넣어주었습니다

seizze reviewed (approved):
안녕하세요 Hailey! 수고 많으셨어요~
전체적으로 꼼꼼하고 신경을 많이 쓴 부분이 보입니다. 👏👏

eunseo5355 committed:
🐛 [Fix] Text View의 Background 색상 오류 수정

- 가로 사이즈가 regular → compact로 바뀔때 그레이로 바뀌지 않았던 문제 수정

eunseo5355 committed:
🐛 [Fix] Text View 스크롤, 수정 관련 오류 수정

- 다른 메모로 넘어갈때 스크롤이 초기화되지 않던 문제 해결
- 현재 메모에서 수정 중이었다가 다른 메모로 넘어갈때 계속 수정 중으로 뜨는 문제 해결

eunseo5355 committed:
♻️ [Refactor] Step 1 피드백 반영

- delegate 패턴 구현할때 첫 번째 프로퍼티로 자기 자신을 넘겨주도록 수정
- DataFormatter를 유틸리티 형식으로 수정
- 텍스트뷰와 뷰모델 사이의 바인딩 코드 추가

eunseo5355 commented:
@seizze 

하이디 안녕하세요🤗
주말은 잘 보내셨나요?ㅎㅎ

피드백에 대한 답변은 위에 달아놓았습니다..!
이번 PR에 대한 피드백은 다음 스텝하면서 수정해볼게요🙃

항상 꼼꼼한 리뷰 감사합니다🙇🏻‍♀️

eunseo5355 committed:
🐛 [Fix] 잘못된 프로퍼티명으로 호출한 오류 수정

seizze commented on CloudNotes/CloudNotes/ViewController/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(true)
+        detailNoteTextView.isEditable = true
+        detailNoteTextView.contentOffset = CGPoint(x: -20, y: -20)
+        configureTextView()
+    }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(true)
+        detailNoteTextView.isEditable = false
+    }
```
super를 호출해 준 건 좋은데 왜 true로 전달했을까요? 이 화면은 항상 animated 되어야 하나요? ㅎㅎ

eunseo5355 replied:
생각해보니 굳이 animated를 해줄 필요가 없었네요..ㅎㅎ 바로 수정했습니다!

seizze commented on CloudNotes/CloudNotes/ViewModel/NoteViewModel.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+struct NoteViewModel {
```
다른 곳에서 이 타입을 관찰하기 위해 NoteViewModel이나 [NoteViewModel]에 대한 Observable을 선언하고 있는데, 뷰모델을 관찰한다고 선언하는 것보단 모델을 관찰한다고 선언하는게 더 자연스러운 것 같습니다.
그리고 여기 타입을 보니, 뷰모델보단 모델에 가까운 것 같네요 ㅎㅎ NoteViewModel을 따로 만들지 말고 Note를 바로 사용해서 Observable로 만드는 건 어떨까요? 다른 의견 있으시면 편하게 말씀 주세요 ☺️
Date 포맷팅 기능도, 노트 리스트 화면에서만 사용하니까 NoteListViewModel에서 갖고 있는게 더 좋을 것 같은데 어떻게 생각하시나요?

eunseo5355 replied:
제가 그 생각을 못했네요..!! 하이디가 말씀해주신 것처럼 모델을 관찰하는게 더 자연스러운 것 같아요
그리고 Date 포맷팅 기능도 NoteListViewModel이 갖고있는게 더 나아보이네요 👍

seizze commented on CloudNotes/CloudNotes/ViewController/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        super.viewWillAppear(true)
+        detailNoteTextView.isEditable = true
+        detailNoteTextView.contentOffset = CGPoint(x: -20, y: -20)
+        configureTextView()
```
음 바인드 시점에 configureTextView가 한번 호출돼서 데이터가 세팅되지 않나요? ㅎㅎ 여기서 또 호출해주지 않아도 될 것 같아요. 그리고 MVVM에서는 이런 식으로 뷰에 데이터를 세팅하는 코드를 직접 호출하는게 아니라, 데이터가 변경되면 바인딩된 코드가 동작해 업데이트되도록 해야합니다.

eunseo5355 replied:
제가 이 코드를 작성하기 전에는 viewDidLoad에서 작성했던 noteTextViewModel의 바인딩 코드가 셀을 눌렀을때 해당 프로퍼티가 세팅되면서 호출될 거라고 예상을 했지만 실제로는 화면에 아무것도 뜨지않았습니다😢 
처음 로드할때 noteTextViewModel이 nil이라서 그런걸까요..??
이 문제를 해결하기 위해 위처럼 configureTextView() 메소드를 호출시켜줬는데 다른 해결책으로는 viewDidLoad에서 호출했던 아래의 코드를 viewWillAppear에서 호출되도록 옮기는 방법을 생각해 보았습니다..!
하이디는 어떻게 생각하시나요?
```Swift
noteTextViewModel?.note.bind { [weak self] _ in
   DispatchQueue.main.async {
      self?.configureTextView()
   }
}
```

seizze replied:
> 제가 이 코드를 작성하기 전에는 viewDidLoad에서 작성했던 noteTextViewModel의 바인딩 코드가 셀을 눌렀을때 해당 프로퍼티가 세팅되면서 호출될 거라고 예상을 했지만 실제로는 화면에 아무것도 뜨지않았습니다😢
처음 로드할때 noteTextViewModel이 nil이라서 그런걸까요..??

확인해 봤는데, 스플릿뷰에서 같은 인스턴스를 재사용하고 있어서 viewDidLoad가 최초 한 번만 불리기 때문이네요~ 해결 방법은 여러가지입니다. 관련 클래스들의 인스턴스가 생성되고 뷰모델이 세팅되는 시점, viewDidLoad/viewWillAppear가 불리는 시점들의 순서가 어떻게 되는지 살펴보면 해결 방법을 찾을 수 있을거에요 😎 이참에 breakpoint로 디버깅하는 방법에도 익숙해지면 좋을 것 같습니다.

> 이 문제를 해결하기 위해 위처럼 configureTextView() 메소드를 호출시켜줬는데 다른 해결책으로는 viewDidLoad에서 호출했던 아래의 코드를 viewWillAppear에서 호출되도록 옮기는 방법을 생각해 보았습니다..!
하이디는 어떻게 생각하시나요?

viewWillAppear는 같은 ViewController 인스턴스에서 여러 번 호출될 수 있습니다. 따라서 바인딩시키는 코드가 여러 번 호출 될 수 있습니다. 지금은 didSet과 클로저를 이용하고 있어서 문제없겠지만, 다른 바인딩 방식을 사용했을 때도 문제가 없을지 잘 모르겠네요 ㅜㅜ 바인딩은 한 번만 호출되도록 하는 게 좋을 것 같아요.

seizze reviewed (approved):
주말까지 열정 💪💪
다음 스텝도 화이팅입니다~!
